### PR TITLE
Making dask fields from pattern known categoricals

### LINKED ISSUE #178: In CSV plugin, `path_as_pattern` field columns should be known categories
In the path column on dask the paths are known categories, which is why they are parse-able before the data are read. It would be good if the `path_as_pattern` fields were also known in this same way. It is a little tricky because the mapping from paths to unique pattern-field values is not necessarily 1:1 so we can't just use the same cat.codes. I guess what we should do is declare a new category index with new codes  and field values and then map path codes to field codes. That should work.

### PR DESCRIPTION
**Pull request by jsignell**
Fixes #178 and adds test

### CODE REVIEW

File: intake/source/csv.py [hunk 1]
```diff
@@ -54,12 +54,19 @@ def __init__(self, urlpath, csv_kwargs=None, metadata=None,
...
+
         col = self._dataframe[path_column]
         paths = col.cat.categories
 
         column_by_field = {field:
-            col.cat.codes.map(dict(enumerate(values))).astype("category") for
-            field, values in reverse_formats(self.pattern, paths).items()
+            col.cat.codes.map(dict(enumerate(values))).astype(
+                "category" if not _HAS_CDT else CategoricalDtype(set(values))
+            ) for field, values in reverse_formats(self.pattern, paths).items()
```
Comment (martindurant) [inline intake/source/csv.py]: So the original type loaded by Dask is still string, but we convert to get categoricals at the next stage?

File: intake/source/tests/test_csv.py [hunk 2]
```diff
@@ -98,6 +98,10 @@ def test_read_chunked(sample1_datasource, data_filenames):
 
 
 def check_read_pattern_output(source):
+    da = source.to_dask()
+    assert da.num.cat.known is True
```
Comment (martindurant) [inline intake/source/tests/test_csv.py]: Presumably we can check the specific values of `da.num.cat` (the metadata), and also grab a partition and (`.get_partition(0).compute()`) and check that it also has the right categories.

File: intake/source/csv.py [hunk 3]
```diff
@@ -54,12 +54,19 @@ def __init__(self, urlpath, csv_kwargs=None, metadata=None,
     def _set_pattern_columns(self, path_column):
         """Get a column of values for each field in pattern
         """
+        try:
+            from pandas.api.types import CategoricalDtype
+            _HAS_CDT = True
```
Comment (martindurant) [inline intake/source/csv.py]: This is a Pandas version thing? Maybe nice to have a clarifying comment.

Comment (jsignell) [reply intake/source/csv.py]: Right it is >=0.21.0 https://pandas.pydata.org/pandas-docs/stable/generated/pandas.api.types.CategoricalDtype.html#pandas.api.types.CategoricalDtype

File: intake/source/tests/test_csv.py (hunk 2 above)
Comment (jsignell) [reply intake/source/tests/test_csv.py]: yeah we can. We check all that below though right?

File: intake/source/csv.py (hunk 1 above)
Comment (jsignell) [reply intake/source/csv.py]: no sure I understand the distinction.

File: intake/source/tests/test_csv.py (hunk 2 above)
Comment (martindurant) [reply intake/source/tests/test_csv.py]: No, below we effectively do `.compute()` and then check everything on the resulting dataframe. I'm suggesting we should also check the `df` metadata before compute, and the contents of one of the partitions. It should be the same, of course.

File: intake/source/csv.py (hunk 1 above)
Comment (martindurant) [reply intake/source/csv.py]: I think I was misunderstanding what happens here. Is the following right:
upon load, dask applies the filename pieces to each chunk, to create categorical columns which always have just a  single label within each partition; then in this step we change the labels to real values based on the parsing. Is that right? In that case, everything is fine.

**Comment by martindurant** [comment]
Closes #178

File: intake/source/csv.py (hunk 3 above)
Comment (jsignell) [reply intake/source/csv.py]: fixed

File: intake/source/tests/test_csv.py (hunk 2 above)
Comment (jsignell) [reply intake/source/tests/test_csv.py]: done

**Comment by jsignell** [comment]
Ok I think this is ready

**Pull request merged**